pkg/controllers/infracluster/vsphere.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain MAPI ProviderSpec: %w", err)
+	}
+
+	providerSpec := &mapiv1beta1.VSphereMachineProviderSpec{}
+	if err := yaml.Unmarshal(rawProviderSpec, providerSpec); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal MAPI ProviderSpec: %w", err)
+	}
+
+	return providerSpec, nil
+}
+
+// ensureVSphereSecret ensures the CAPI VSphere credentials secret exists.
+func (r *InfraClusterController) ensureVSphereSecret(ctx context.Context, vsphereServerAddr string) error {
+	vSphereSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+		},
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(vSphereSecret), vSphereSecret); err != nil && !cerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get CAPI VSphere credentials secret: %w", err)
+	}
+
+	username, password, err := r.getVSphereCredentials(ctx, vsphereServerAddr)
+	if err != nil {
+		return fmt.Errorf("unable to get VSphere credentials: %w", err)
+	}
+
+	vSphereSecret.StringData = map[string]string{
+		"username": username,
+		"password": password,
+	}
+
+	if err := r.Client.Create(ctx, vSphereSecret); err != nil && !cerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("unable to create CAPI VSphere credentials secret: %w", err)
+	}
+
+	return nil
+}
+
+// getVSphereCredentials obtains the VSphere credentials from the well-known credentials secret.
+func (r *InfraClusterController) getVSphereCredentials(ctx context.Context, vsphereServerAddr string) (string, string, error) {
+	vSphereCredentialsSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: kubeSystemNamespace,
+		Name:      vSphereCredentialsName,
+	}, vSphereCredentialsSecret); err != nil {
+		return "", "", fmt.Errorf("%w %s/%s: %w", errUnableToFindCredsSecret, kubeSystemNamespace, vSphereCredentialsName, err)
+	}
+
+	username, ok := vSphereCredentialsSecret.Data[fmt.Sprintf("%s.username", vsphereServerAddr)]
+	if !ok {
+		return "", "", fmt.Errorf("%w %s/%s", errUnableToFindUsernameVSphereCredsSecret, kubeSystemNamespace, vSphereCredentialsName)
+	}
+
+	password, ok := vSphereCredentialsSecret.Data[fmt.Sprintf("%s.password", vsphereServerAddr)]
+	if !ok {
+		return "", "", fmt.Errorf("%w %s/%s", errUnableToFindPasswordVSphereCredsSecret, kubeSystemNamespace, vSphereCredentialsName)
+	}
+
+	return string(username), string(password), nil
+}
+
+// getVSphereServerAddr obtains the VSphere Server address.
+func (r *InfraClusterController) getVSphereServerAddr(ctx context.Context) (string, error) {
+	if r.Infra.Spec.PlatformSpec.VSphere == nil || len(r.Infra.Spec.PlatformSpec.VSphere.VCenters) == 0 {
+		// Devise VSphere server addr via MAPI providerSpec.
+		machineSpec, err := getVSphereMAPIProviderSpec(ctx, r.Client)
+		if err != nil || machineSpec == nil {
+			return "", errUnableToGetVSphereMAPIProviderSpec
```

Comment:
Will this error ever be reported? It looks like this is the only possible error that could happen at L49 in this file, and that function returns a different error without wrapping.

Reply:
Changed the logic around to make `machineSpec == nil` unnecessary and always wrap the err